### PR TITLE
[SPARK-8874] [ML] Add missing methods in Word2Vec

### PR DESCRIPTION
**Pull request by MechCoder**
Add missing methods
1. getVectors
2. findSynonyms

to W2Vec scala and python API

@mengxr

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala [hunk 1]
```diff
@@ -146,6 +151,55 @@ class Word2VecModel private[ml] (
     wordVectors: feature.Word2VecModel)
   extends Model[Word2VecModel] with Word2VecBase {
 
+
+  /**
+   * Return a map of every word to its Vector representation.
+   */
+  val getVectors: Map[String, Array[Float]] = wordVectors.getVectors
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: @mengxr I would like to return this as a dataframe with each row having two fields, a string and a float array. But I'm unable to understand how to access the SparkContext here.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala [hunk 2]
```diff
@@ -146,6 +151,55 @@ class Word2VecModel private[ml] (
...
+    // necessary.
+    val vectorsList = vectors.toList.map(vec => Vectors.dense(vec.map(_.toDouble)))
+    List(words.toList, vectorsList).map(_.asJava.asInstanceOf[Object]).asJava
+  }
+
+  /**
+   * Find "num" no. words closest to similarity to the given word.
+   * Returns a coupled array with the second element giving the similarity measure.
+   */
+  def findSynonyms(word: String, num: Int): Array[(String, Double)] = {
```
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: similarly here.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala (hunk 1 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: So its my understanding, from my reading of the code, it isn't possible to access the Spark context here. If we look at similar methods in other models (like predict in classifier), they can't access the Spark Context, its only when we are applying fit/transform and we have a dataframe passed into us that we can access the SparkContext.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala (hunk 2 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: Similar to the comment above, I don't think we can access the Spark Context here, although perhaps one option (and I don't really know if its a good one) would be to do something similar to the predict method and allow take either a dataframe or rdd of multiple inputs and return a dataframe/rdd of the synonyms.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala (hunk 1 above)
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: Thanks for clarifying. I just thought I was being ignorant.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala (hunk 2 above)
Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: Hmm. I had a similar thought in mind. If I do that I could remove the "java-friendly" code below.

@mengxr thoughts?

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala (hunk 1 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: Actually, we can call `SparkContext.getOrCreate()` to get the active `SparkContext`.

Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: oh my bad, I didn't see that new functionality.

**Comment by MechCoder** [comment]
@mengxr I added support only for the scala code right now. We can move the python wrappers to another PR.

**Comment by MechCoder** [comment]
ping @mengxr

**Comment by jkbradley** [comment]
I'm curious: Did you discuss with @mengxr about whether these methods should return the same types as the spark.mllib Scala API vs. DataFrames?

**Comment by MechCoder** [comment]
I remember discussing somewhere (I think offline on chat). However, I think it would be neater to return DataFrames and would comply with spark.ml in general. What is your opinion?

**Comment by jkbradley** [comment]
OK, if it was discussed, then it's fine with me.  After all, it's more Java-friendly.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala [hunk 3]
```diff
@@ -146,6 +148,40 @@ class Word2VecModel private[ml] (
...
+  val getVectors: DataFrame = {
+    val sc = SparkContext.getOrCreate()
+    val sqlContext = SQLContext.getOrCreate(sc)
+    import sqlContext.implicits._
+    val wordVec = wordVectors.getVectors.mapValues(vec => Vectors.dense(vec.map(_.toDouble)))
+    sc.parallelize(wordVec.toSeq).toDF("word", "vector")
+  }
+
+  /**
+   * Find "num" no. words closest in similarity to the given word.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: "no." --> "number of"

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala [hunk 4]
```diff
@@ -146,6 +148,40 @@ class Word2VecModel private[ml] (
...
+   * Returns a dataframe with the words and the cosine similarities between the
+   * synonyms and the given word.
+   */
+  def findSynonyms(word: String, num: Int): DataFrame = {
+    findSynonyms(wordVectors.transform(word), num)
+  }
+
+  /**
+   * Find "num" no. words closest to similarity to the vector representation of the word.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: "Find "num" number of words closest in similarity to the given vector representation of a word."

File: mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala [hunk 5]
```diff
@@ -146,6 +148,40 @@ class Word2VecModel private[ml] (
...
+   * Returns a dataframe with the words and the cosine similarities between the
+   * synonyms and the given word.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala]: "given word" --> "given word vector"

File: mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala [hunk 6]
```diff
@@ -67,5 +67,67 @@ class Word2VecSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+  test("getVectors") {
+
+    val sqlContext = new SQLContext(sc)
+    import sqlContext.implicits._
+
+    val sentence = "a b " * 100 + "a c " * 10
+    val doc = sc.parallelize(Seq(sentence, sentence)).map(line => line.split(" "))
+
+    val codes = Map(
+      "a" -> Array(-0.2811822295188904, -0.6356269121170044, -0.3020961284637451),
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala]: Were these values computed using another library?  If so, it'd be nice to document that.

Also, can you avoid having multiple copies of them in the unit test, and instead put them in a static object instead?

File: mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala [hunk 7]
```diff
@@ -67,5 +67,67 @@ class Word2VecSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+    val docDF = doc.zip(doc).toDF("text", "alsotext")
+
+    val model = new Word2Vec()
+      .setVectorSize(3)
+      .setInputCol("text")
+      .setOutputCol("result")
+      .setSeed(42L)
+      .fit(docDF)
+
+    val expectedSimilarity = Array(0.2789285076917586, -0.6336972059851644)
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala]: How were these computed?

File: mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala (hunk 6 above)
Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala]: Ah, I think I just used the `Word2Vec` in `mllib` . Should I test it with another library or is this sufficient?

Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala]: I guess it's fine.  We need to beef up the mllib tests at some point.  I'll make a JIRA for that.

**Comment by jkbradley** [comment]
LGTM.  Merging with master and branch-1.5
Thanks!

**Pull request closed without merging**